pandas/core/dtypes/cast.py

```diff
@@ -1026,3 +1027,28 @@ def find_common_type(types):
             return np.object
 
     return np.find_common_type(types, [])
+
+
+def maybe_cast_to_integer(arr, dtype):
+    """
+    Find a common data type among the given dtypes.
+
+    Parameters
+    ----------
+    arr : array
+    dtype : dtype
+
+    Returns
+    -------
+    integer or unsigned integer array (or raise if the dtype is incompatible)
+
+    """
+
+    if is_unsigned_integer_dtype(dtype) and (np.asarray(arr) < 0).any():
+        raise OverflowError("Trying to coerce negative values to negative "
+                            "integers")
+    elif is_integer_dtype(dtype) and is_float_dtype(np.asarray(arr)):
+        if ((np.asarray(arr) % np.asarray(arr).astype(int)) > 0).any():
```

Comment:
for floats, the check is
``(arr == arr.astype(dtype)).all()``

its not an ``OverFlowError``, rather a ``ValueErorr`` here